### PR TITLE
c8d: Set the process user on exec

### PR DESCRIPTION
**Pull request by rumpl**
**- What I did**

Set the process user on exec

This change makes is possible to run `docker exec -u <UID> ...` when the containerd integration is activated.

- Upstreams https://github.com/rumpl/moby/pull/86
- Upstreams https://github.com/rumpl/moby/pull/92

**- How I did it**

We load the container spec and change its user and then set the exec process user.

**- How to verify it**

```console
$ docker run -d --rm -it ubuntu /bin/bash
f986fab0f34af0d235a10a7df32199f54d79a4da4d1984b85bcab5fd5fffbe27
$ docker exec -u 1000 f986fab0f34af0d235a10a7df32199f54d79a4da4d1984b85bcab5fd5fffbe27 id
uid=1000 gid=0(root) groups=0(root)
```
**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by rumpl** [comment]
Converting to draft for now, this one needs this too https://github.com/rumpl/moby/pull/92

**Comment by rumpl** [comment]
I opened a PR in containerd https://github.com/containerd/containerd/pull/8351

**Comment by rumpl** [comment]
The readonly mount in containerd was merged in master and 1.6 and 1.7 release branches, I updated the dependency to https://github.com/containerd/containerd/commit/c0efc63d3907e530740a71ca8dc5e559a165dbdd, exec-ing into a container with a user now works and doesn't put the mounts into an undefined state.

File: daemon/exec_linux.go [hunk 1]
```diff
@@ -3,27 +3,77 @@ package daemon // import "github.com/docker/docker/daemon"
...
+}
+
 func (daemon *Daemon) execSetPlatformOpt(ctx context.Context, ec *container.ExecConfig, p *specs.Process) error {
 	if len(ec.User) > 0 {
 		var err error
-		p.User, err = getUser(ec.Container, ec.User)
-		if err != nil {
-			return err
+		if daemon.UsesSnapshotter() {
+			p.User, err = getUserFromContainerd(ctx, daemon.containerdCli, ec)
```
Comment (vvoland) [inline daemon/exec_linux.go:57]: Could this one also work for `!daemon.UsesSnapshotter()` branch?

**Review by vvoland — APPROVED** [review]

File: daemon/exec_linux.go (hunk 1 above)
Comment (rumpl) [reply daemon/exec_linux.go:57]: It can't work no, the mounts of the container are handled by the graph drivers

Comment (rumpl) [reply daemon/exec_linux.go:57]: Ok just tested and indeed this can work for both cases, will update

Comment (rumpl) [reply daemon/exec_linux.go:57]: I'd rather we didn't actually, we can do it in the future once we are sure that the two ways of getting the users and groups are equivalent, I can add a TODO though

Comment (vvoland) [reply daemon/exec_linux.go:57]: Yeah it might be better not to touch the non-snapshotter case. I was just asking out of curiosity 😄

**Review by thaJeztah — APPROVED** [review]
LGTM (I guess technically the commits should have been in the reverse order, but the first one "builds", so I guess it's fine)

**Pull request merged**